File: Scala/compoundlab/src/main/scala/GooseAdapter.scala
class GooseAdapter(goose : Goose, observable: Option[IQuackableObservable] = None) extends IQuackable
{
	private val name = "GooseAdapter"
	private val _goose : Goose = goose
	private val _observable : IQuackableObservable = observable getOrElse new Observable(this)


	override def registerObserver(observer : IObserver) : Unit = {
		this._observable.registerObserver(observer)
	}

	override def notifyObservers: String = {
		this._observable.notifyObservers
	}
	override def quack : String = {
		notifyObservers
		this._goose.honk
	}

	override def getName: String = {
		this.name
	}
}

File: Java/compoundlab/src/main/java/com/bbd/abstractFactoryPattern/GooseAdapter.java
package com.bbd.abstractFactoryPattern;

public class GooseAdapter implements IQuackable
{
	private Goose _goose;
	private IQuackObservable _observable;
	private String _name = "Goose";

	GooseAdapter(Goose goose)
	{
		this._goose = goose;
		this._observable = new Observable(this);
	}

	GooseAdapter(Goose goose, IQuackObservable observable)
	{
		this._goose = goose;
		this._observable = observable;
	}

	public void registerObserver(IObserver observer)
	{
		this._observable.registerObserver(observer);
	}

	public String notifyObservers()
	{
		return (this._observable.notifyObservers());
	}

	public String quack()
	{
		notifyObservers();
		return (this._goose.honk());
	}

	@Override
	public String getName() { return  this._name;}
}
